### PR TITLE
Use proxy types for package private properties

### PR DESCRIPTION
**Pull request by MaxDesiatov**
I think that while public functions giving access to private properties does the job, they pollute the global namespace of `TokamakCore`, while for every property you need to create a separate function. These functions read a bit weird at their place of use, so I'm wondering whether the Proxy pattern could work better here. While it's more verbose, I find it a bit more readable, while also adding some sort of structure here for grouping multiple properties if needed.

CC @j-f1, this is something you might want to use in #125.

### CODE REVIEW

**Comment by ie-ahm-robox** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/Button.swift#L21** - FIXMEs should be resolved (this should be internal). (`todo`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/SecureField.swift#L21** - FIXMEs should be resolved (this should be internal). (`todo`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/TextField.swift#L21** - FIXMEs should be resolved (this should be internal). (`todo`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/TextField.swift#L43** - FIXMEs should be resolved (implement this method, which u...). (`todo`)
  </td>
    </tr>
  
<tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakDOM/Views/Text.swift#L149** - Line should be 100 characters or less: currently 103 characters (`line_length`)
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/swift">Danger Swift</a> against a65c37cbba667066c333590471c72f94a89fe19a
</p>

**Comment by carson-katri** [comment]
Should there be a `_Proxy` protocol to save on some boilerplate?

```swift
protocol _Proxy {
  associatedtype Subject : View
  var subject: Subject { get }
  init(_ subject: Subject)
}

extension _Proxy {
  init(_ subject: Subject) { self.subject = subject }
}
```

**Comment by j-f1** [comment]
I’m not exactly sure how this works for things like styles where we want to be able to directly tag the view objects before rendering them.

**Comment by MaxDesiatov** [comment]
IDK, as far as I know, at least according to compiler engineers on Swift Forums a plain struct with a single field is equivalent to using the value of that single field directly. I also hope that when inlining is applied, there really ырщгдв be no noticeable performance overhead with proxies. On the other hand, as soon as you introduce protocols and extensions, it adds all the runtime metadata cruft, which could have some impact in these hot code paths. 

I think we could apply some meta-programming magic to reduce the amount of boilerplate at some point, probably similar to [Sourcery](https://github.com/krzysztofzablocki/Sourcery) or GYB, but without the textual substitutions unsafety. I think I got a few ideas for that inspired by [Rust macros](https://doc.rust-lang.org/book/ch19-06-macros.html) and [SwiftSyntax](https://github.com/apple/swift-syntax), but nothing ready to be shown yet...

File: Sources/TokamakCore/Views/SecureField.swift [hunk 1]
```diff
@@ -34,10 +32,17 @@ extension SecureField where Label == Text {
   ) where S: StringProtocol {
     label = Text(title)
     textBinding = text.projectedValue
-    commitAction = onCommit
+    self.onCommit = onCommit
```
Comment (j-f1) [inline Sources/TokamakCore/Views/SecureField.swift:35]: Is the convention in `init` to specify `self.` or not?

File: Sources/TokamakCore/Views/TupleView.swift [hunk 2]
```diff
@@ -22,65 +22,66 @@ public struct TupleView<T>: View {
...
+    _children = children
   }
 
   init<T1: View, T2: View>(_ v1: T1, _ v2: T2) where T == (T1, T2) {
     value = (v1, v2)
-    children = [AnyView(v1), AnyView(v2)]
+    _children = [AnyView(v1), AnyView(v2)]
   }
 
   init<T1: View, T2: View, T3: View>(_ v1: T1, _ v2: T2, _ v3: T3) where T == (T1, T2, T3) {
```
Comment (ie-ahm-robox) [inline Sources/TokamakCore/Views/TupleView.swift:42]: - :no_entry_sign: Tuples should have at most 2 members. (`large_tuple`)

File: Sources/TokamakCore/Views/TextField.swift [hunk 3]
```diff
@@ -36,8 +34,8 @@ extension TextField where Label == Text {
   ) where S: StringProtocol {
     label = Text(title)
     textBinding = text.projectedValue
-    editingChangedAction = onEditingChanged
-    commitAction = onCommit
+    self.onEditingChanged = onEditingChanged
+    self.onCommit = onCommit
   }
 
   // FIXME: implement this method, which uses a Formatter to control the value of the TextField
```
Comment (ie-ahm-robox) [inline Sources/TokamakCore/Views/TextField.swift:41]: - :no_entry_sign: FIXMEs should be resolved (implement this method, which u...). (`todo`)

**Comment by ie-ahm-robox** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakDOM/Views/Text.swift#L149** - Line should be 100 characters or less: currently 103 characters (`line_length`)
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/swift">Danger Swift</a> against a19d367182e03be5fe47c23829d25f5ab79d23cb
</p>

File: Sources/TokamakCore/Views/Button.swift [hunk 4]
```diff
@@ -45,6 +44,12 @@ extension Button: ParentView {
...
-public func buttonLabel(_ button: Button<Text>) -> String {
-  button.label.content
+/// This is a helper class that works around absence of "package private" access control in Swift
+public struct _ButtonProxy {
+  public let subject: Button<Text>
+
+  public init(_ subject: Button<Text>) { self.subject = subject }
+
+  public var label: _TextProxy { _TextProxy(subject.label) }
+  public var action: () -> () { subject.action }
```
Comment (j-f1) [inline Sources/TokamakCore/Views/Button.swift:54]: ok this makes a lot more sense! So it allows anyone with access to `_ButtonProxy` to read private properties of a `Button`.

File: Sources/TokamakDOM/Views/Button.swift [hunk 5]
```diff
@@ -21,8 +21,8 @@ public typealias Button = TokamakCore.Button
 
 extension Button: ViewDeferredToRenderer where Label == Text {
   public var deferredBody: AnyView {
-    AnyView(HTML("button", listeners: ["click": { _ in action() }]) {
-      Text(buttonLabel(self))
+    AnyView(HTML("button", listeners: ["click": { _ in _ButtonProxy(self).action() }]) {
```
Comment (j-f1) [inline Sources/TokamakDOM/Views/Button.swift:24]: Is there a reason this isn’t
```suggestion
    AnyView(HTML("button", listeners: ["click": _ButtonProxy(self).action]) {
```

File: Sources/TokamakCore/Views/TextField.swift [hunk 6]
```diff
@@ -17,11 +17,9 @@
...
 public struct TextField<Label>: View where Label: View {
   let label: Label
-
-  // FIXME: this should be internal
-  public let textBinding: Binding<String>
-  public let editingChangedAction: (Bool) -> ()
-  public let commitAction: () -> ()
+  let textBinding: Binding<String>
+  let onEditingChanged: (Bool) -> ()
+  let onCommit: () -> ()
```
Comment (MaxDesiatov) [inline Sources/TokamakCore/Views/TextField.swift:22]: @j-f1 you could the property here as you did in #125, but as internal, then add a corresponding computed property for it in the proxy, does that make sense?

File: Sources/TokamakCore/Views/TupleView.swift (hunk 2 above)
Comment (ie-ahm-robox) [inline Sources/TokamakCore/Views/TupleView.swift:42]: - :no_entry_sign: Tuples should have at most 2 members. (`large_tuple`)

**Comment by ie-ahm-robox** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/TextField.swift#L41** - FIXMEs should be resolved (implement this method, which u...). (`todo`)
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/swift">Danger Swift</a> against e297abc20f43053e2d9b0860b44576ad8dd193fd
</p>

File: Sources/TokamakCore/Views/TupleView.swift (hunk 2 above)
Comment (j-f1) [reply Sources/TokamakCore/Views/TupleView.swift:42]: This can be fixed by moving the comment below up to above this one.

Comment (MaxDesiatov) [reply Sources/TokamakCore/Views/TupleView.swift:42]: Great shout, not sure how I missed that

File: Sources/TokamakCore/Views/Button.swift (hunk 4 above)
Comment (MaxDesiatov) [reply Sources/TokamakCore/Views/Button.swift:54]: Yes, same as stand-alone functions did, but now grouped in a proxy wrapper type.

File: Sources/TokamakCore/Views/SecureField.swift (hunk 1 above)
Comment (MaxDesiatov) [reply Sources/TokamakCore/Views/SecureField.swift:35]: Yes, I personally prefer stored properties to have the same names as arguments, which forces you to use `self`. I didn't mind it as much in your PR, but thought I could make it more consistent here with the names of computed proxy properties.

File: Sources/TokamakDOM/Views/Button.swift (hunk 5 above)
Comment (MaxDesiatov) [reply Sources/TokamakDOM/Views/Button.swift:24]: None other than laziness 😄

File: Sources/TokamakCore/Views/TupleView.swift [hunk 7]
```diff
@@ -22,65 +22,66 @@ public struct TupleView<T>: View {
...
+    _children = children
   }
 
   init<T1: View, T2: View>(_ v1: T1, _ v2: T2) where T == (T1, T2) {
     value = (v1, v2)
-    children = [AnyView(v1), AnyView(v2)]
+    _children = [AnyView(v1), AnyView(v2)]
   }
 
+  // swiftlint:disable large_tuple
```
Comment (ie-ahm-robox) [inline Sources/TokamakCore/Views/TupleView.swift:42]: - :no_entry_sign: Tuples should have at most 2 members. (`large_tuple`)

Comment (ie-ahm-robox) [inline Sources/TokamakCore/Views/TupleView.swift:42]: - :no_entry_sign: Tuples should have at most 2 members. (`large_tuple`)

**Comment by ie-ahm-robox** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/TextField.swift#L41** - FIXMEs should be resolved (implement this method, which u...). (`todo`)
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/swift">Danger Swift</a> against dc792ae14b05cdeb2ab6ee640b7c7a0c80d3f332
</p>

**Comment by ie-ahm-robox** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true">Fails</th>
    </tr>
  </thead>
  <tbody><tr>
      <td>:no_entry_sign:</td>
      <td>

  **Sources/TokamakCore/Views/TextField.swift#L41** - FIXMEs should be resolved (implement this method, which u...). (`todo`)
  </td>
    </tr>
  </tbody>
</table>




<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/swift">Danger Swift</a> against dc792ae14b05cdeb2ab6ee640b7c7a0c80d3f332
</p>

File: Sources/TokamakCore/Views/TextField.swift (hunk 3 above)
Comment (ie-ahm-robox) [inline Sources/TokamakCore/Views/TextField.swift:41]: - :no_entry_sign: FIXMEs should be resolved (implement this method, which u...). (`todo`)

File: Sources/TokamakCore/Views/SecureField.swift (hunk 1 above)
Comment (j-f1) [reply Sources/TokamakCore/Views/SecureField.swift:35]: I’m fine either way — feel free to change it or I could change it in the future.

File: Sources/TokamakDOM/Views/Button.swift (hunk 5 above)
Comment (MaxDesiatov) [reply Sources/TokamakDOM/Views/Button.swift:24]: Turns out another reason is `error: type of expression is ambiguous without more context`.

File: Sources/TokamakCore/Views/TextField.swift (hunk 3 above)
Comment (ie-ahm-robox) [inline Sources/TokamakCore/Views/TextField.swift:41]: - :no_entry_sign: FIXMEs should be resolved (implement this method, which u...). (`todo`)

**Comment by MaxDesiatov** [comment]
@carson-katri this one's ready for review too 🙂

**Pull request merged**